interactor/src/sovereign_forge/sovereign_forge_interactor_main.rs

```diff
@@ -125,79 +148,202 @@ impl SovereignForgeInteract {
 
         self.deploy_chain_factory(
-            sov_forge_address,
-            chain_config_address,
-            header_verifier_address,
-            mvx_esdt_safe_address,
-            fee_market_address,
+            caller.clone(),
+            preferred_chain_id.clone(),
+            sovereign_forge_address.clone(),
+            template_addresses.clone(),
         )
         .await;
-        let chain_factory_address = self.state.current_chain_factory_sc_address().clone();
+        self.register_chain_factory(initial_caller.clone(), shard_id, preferred_chain_id.clone())
+            .await;
 
-        self.deploy_token_handler(chain_factory_address.to_address())
+        self.deploy_token_handler(caller.clone(), preferred_chain_id.clone())
             .await;
+        self.register_token_handler(initial_caller.clone(), shard_id, preferred_chain_id)
+            .await;
+    }
 
-        self.register_token_handler(0).await;
-        self.register_token_handler(1).await;
-        self.register_token_handler(2).await;
-        self.register_token_handler(3).await;
-        self.register_chain_factory(0).await;
-        self.register_chain_factory(1).await;
-        self.register_chain_factory(2).await;
-        self.register_chain_factory(3).await;
+    pub async fn deploy_on_one_shard(
+        &mut self,
+        shard: u32,
+        deploy_cost: BigUint<StaticApi>,
+        optional_esdt_safe_config: OptionalValue<EsdtSafeConfig<StaticApi>>,
+        optional_sov_config: OptionalValue<SovereignConfig<StaticApi>>,
+        fee: Option<FeeStruct<StaticApi>>,
+    ) {
+        let caller = self.get_sovereign_owner_for_shard(shard);
+        let preferred_chain_id = PREFERRED_CHAIN_IDS[shard as usize].to_string();
+        self.deploy_phase_one(
+            caller.clone(),
+            deploy_cost.clone(),
+            Some(preferred_chain_id.clone().into()),
+            optional_sov_config.clone(),
+        )
+        .await;
+        self.deploy_phase_two(caller.clone(), optional_esdt_safe_config.clone())
+            .await;
+        self.deploy_phase_three(caller.clone(), fee.clone()).await;
+        self.deploy_phase_four(caller.clone()).await;
+
+        self.complete_setup_phase(caller.clone()).await;
+        self.check_setup_phase_status(&preferred_chain_id, true)
+            .await;
 
-        self.deploy_phase_one(deploy_cost, Some(chain_id.into()), optional_sov_config)
+        self.update_smart_contracts_addresses_in_state(preferred_chain_id.clone())
             .await;
-        self.deploy_phase_two(optional_esdt_safe_config).await;
-        self.deploy_phase_three(fee).await;
-        self.deploy_phase_four().await;
 
-        self.complete_setup_phase().await;
-        self.check_setup_phase_status(chain_id, true).await;
-        self.update_smart_contracts_addresses_in_state(chain_id)
+        self.deploy_testing_sc(caller.clone(), preferred_chain_id)
```

Comment:
testing-sc is not connected to preferred_chain_id, is just a random contract in the chain, in a random shard